### ISSUE TITLE
Integration losing account password

### ISSUE THREAD

**Issue by AllPurposeBen** [bdr99/pymazda#13]
I didn't notice this before because it takes several days to manifest but twice since I installed this branch, the integration broke because it seems to have forgotten the password for the MazdaConnect account I'm using. As soon as I re-enter it, it works fine.

I can't reproduce on demand, rebooting the host or restarting HA doesn't trigger it. Not sure if it's the integration or HA itself. 

It's happened a couple times since swapping out the test code for 2021.7.2 update.

I'm on core-2021.7.2.

_Originally posted by @AllPurposeBen in https://github.com/bdr99/pymazda/issues/11#issuecomment-878350820_

**Comment by bdr99** [comment]
Can you please add this to your configuration.yaml?
```yaml
logger:
  default: warn
  logs:
    pymazda: debug
```
Then, if the issue happens again, can you check your HA logs for any errors or anything that looks out of the ordinary, and share it here if you find anything?

**Comment by AllPurposeBen** [comment]
I enabled logging last week and promptly forgot about it. I checked today but had already started the update process to Core 2021.7.3. Before the update, it was in the failed state but I didn't get a chance to look at the logs. Post update, it's not failing and logs look normal:
```
2021-07-19 13:15:11 DEBUG (MainThread) [pymazda.connection] Logging in as mazda_api1@allpurposeben.info
2021-07-19 13:15:11 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 13:15:15 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 13:15:16 DEBUG (MainThread) [pymazda.connection] Successfully logged in as mazda_api1@allpurposeben.info
2021-07-19 13:15:16 DEBUG (MainThread) [pymazda.connection] Retrieving encryption keys
2021-07-19 13:15:16 DEBUG (MainThread) [pymazda.connection] Sending POST request to service/checkVersion
2021-07-19 13:15:17 DEBUG (MainThread) [pymazda.connection] Successfully retrieved encryption keys
2021-07-19 13:15:17 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVecBaseInfos/v4
2021-07-19 13:15:18 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getNickName/v4
2021-07-19 13:15:18 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getNickName/v4
2021-07-19 13:15:19 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:15:19 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:18:20 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:18:20 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:21:21 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:21:21 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:24:23 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 13:24:23 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
```

Will check back and see if it fails again or if maybe 2021.7.3 was a factor and it's fixed.

**Comment by AllPurposeBen** [comment]
Checked back and HA isn't warning that it's lost the password but I see this in the logs:
```
2021-07-19 15:34:46 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 15:34:47 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 15:34:47 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 15:34:48 ERROR (MainThread) [homeassistant.components.mazda] Unknown error occurred during Mazda update request: 0, message='Attempt to decode JSON with unexpected mimetype: ', url=URL('https://ptznwbh8.mazda.com/appapi/v1/user/login')
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/components/mazda/__init__.py", line 150, in async_update_data
statuses = await gather_with_concurrency(5, *vehicle_status_tasks)
File "/usr/src/homeassistant/homeassistant/util/async_.py", line 170, in gather_with_concurrency
return await gather(
File "/usr/src/homeassistant/homeassistant/util/async_.py", line 168, in sem_task
return await task
File "/usr/src/homeassistant/homeassistant/components/mazda/__init__.py", line 43, in with_timeout
return await task
File "/usr/local/lib/python3.9/site-packages/pymazda/client.py", line 64, in get_vehicle_status
vehicle_status_response = await self.controller.get_vehicle_status(vehicle_id)
File "/usr/local/lib/python3.9/site-packages/pymazda/controller.py", line 31, in get_vehicle_status
response = await self.connection.api_request("POST", "remoteServices/getVehicleStatus/v4", body_dict=post_body, needs_keys=True, needs_auth=True)
File "/usr/local/lib/python3.9/site-packages/pymazda/connection.py", line 158, in api_request
return await self.__api_request_retry(method, uri, query_dict, body_dict, needs_keys, needs_auth, num_retries=0)
File "/usr/local/lib/python3.9/site-packages/pymazda/connection.py", line 167, in __api_request_retry
await self.__ensure_token_is_valid()
File "/usr/local/lib/python3.9/site-packages/pymazda/connection.py", line 262, in __ensure_token_is_valid
await self.login()
File "/usr/local/lib/python3.9/site-packages/pymazda/connection.py", line 312, in login
login_response_json = await login_response.json()
File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1097, in json
raise ContentTypeError(
aiohttp.client_exceptions.ContentTypeError: 0, message='Attempt to decode JSON with unexpected mimetype: ', url=URL('https://ptznwbh8.mazda.com/appapi/v1/user/login')
2021-07-19 15:34:48 ERROR (MainThread) [homeassistant.components.mazda] Error fetching mazda data: 0, message='Attempt to decode JSON with unexpected mimetype: ', url=URL('https://ptznwbh8.mazda.com/appapi/v1/user/login')
2021-07-19 15:34:48 DEBUG (MainThread) [pymazda.connection] Successfully logged in as redacted_mazda_api_username@domain.tld
```

And also:
```
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Server reports request was not encrypted properly. Retrieving new encryption keys.
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Retrieving encryption keys
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Sending POST request to service/checkVersion
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Server reports request was not encrypted properly. Retrieving new encryption keys.
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Retrieving encryption keys
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Sending POST request to service/checkVersion
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Successfully retrieved encryption keys
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4 - attempt #2
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Successfully retrieved encryption keys
2021-07-19 16:56:15 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4 - attempt #2
2021-07-19 16:59:16 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 16:59:16 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
```

**Comment by AllPurposeBen** [comment]
Checked today and it did show the integration stopped and needing configuration. Here's what I saw in the log:
```
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Access token is expired. Fetching a new one.
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Logging in as redacted_mazda_api_username@domain.tld
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] No access token present. Logging in.
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Logging in as redacted_mazda_api_username@domain.tld
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 23:37:56 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 23:37:56 WARNING (SyncWorker_5) [homeassistant.components.ring] Error fetching Ring history data for device 4c24984c27cd: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2021-07-19 23:37:56 DEBUG (MainThread) [pymazda.connection] Successfully logged in as redacted_mazda_api_username@domain.tld
2021-07-19 23:37:56 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Login failed due to invalid email or password
2021-07-19 23:37:57 ERROR (MainThread) [homeassistant.components.mazda] Authentication failed while fetching mazda data: Not authenticated with Mazda API
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Server reports access token was expired. Retrieving new access token.
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Logging in as redacted_mazda_api_username@domain.tld
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 23:37:57 DEBUG (MainThread) [pymazda.connection] Successfully logged in as redacted_mazda_api_username@domain.tld
```

Restarting core cleared the issue and it's running again. I did notice some errors from other integrations (I left a ring error in the above log snip) and will investigate if maybe there's a large issue in my HA instance. Pease advise if the above logging would indicate that.

**Comment by bdr99** [comment]
Thanks for sharing those logs. FYI, the message `Server reports request was not encrypted properly` is completely normal. There is an encryption key which changes every so often, and that message just means it is fetching the latest encryption key from the Mazda API. `Server reports access token was expired. Retrieving new access token.` is also normal. The access token expires after some time, requiring the library to complete the login process again to fetch a new one.

The other errors, however, are not normal. `Attempt to decode JSON with unexpected mimetype` seems to indicate that the response from the Mazda API could not be parsed as valid JSON, and `Login failed due to invalid email or password` would usually mean that your email address and password are wrong. These are strange, but what I find the most strange is this section:
```
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-19 23:37:55 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-19 23:37:56 DEBUG (MainThread) [pymazda.connection] Sending login request
```
This tells me that it is trying to log into the account twice at the same time, for some reason. For reference, here is a section from my own HA logs that shows what a typical access token expiration should look like (note that `Sending login request` only appears once):
```
2021-07-20 19:43:18 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-20 19:46:19 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-20 19:49:20 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Access token is expired. Fetching a new one.
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Logging in as myemail@domain.tld
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Retrieving public key to encrypt password
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Sending login request
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Successfully logged in as myemail@domain.tld
2021-07-20 19:52:21 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-20 19:55:23 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
2021-07-20 19:58:24 DEBUG (MainThread) [pymazda.connection] Sending POST request to remoteServices/getVehicleStatus/v4
```
Can you think of anything unique about your setup that would cause it to spawn two instances of the Mazda integration? When you go to Configuration -> Integrations, do you see two accounts listed under Mazda?

**Comment by AllPurposeBen** [comment]
> Can you think of anything unique about your setup that would cause it to spawn two instances of the Mazda integration? When you go to Configuration -> Integrations, do you see two accounts listed under Mazda?

I only see one. Unhid ignored and disabled and confirmed only one. 

I've been looking at the `ha` command line tool to see if there's a way to see running instances of an integration but I'm not seeing how and not sure how they are scheduled/daemonized.

The only non-standard thing I can think of I've done related to this integration was running the test code from the last release. I confirmed that it's no longer in `config/custom_components`.

**Comment by bdr99** [comment]
Can you check the file `.storage/core.config_entries` in your HA config directory? In the `entries` array, how many entries are there where the `domain` is `"mazda"`?

**Comment by AllPurposeBen** [comment]
Just the one entry:
```
{
	"entry_id": "5755ac37d22eeff3497f31ca75ccda72",
	"version": 1,
	"domain": "mazda",
	"title": "Mazda Connect",
	"data": {
		"email": "redacted_mazda_api_username@domain.tld",
		"region": "MNAO",
		"password": "password"
	},
	"options": {},
	"pref_disable_new_entities": false,
	"pref_disable_polling": false,
	"source": "user",
	"unique_id": "redacted_mazda_api_username@domain.tld",
	"disabled_by": null
},
```

**Comment by AllPurposeBen** [comment]
Worth noting, I have 2 vehicles in my Mazda account. I'm not sure it's that's a factor and why there are 2 calls. I see data for both in `core.device_registry`.

**Comment by bdr99** [comment]
Ah, that's a good point. I think that is probably the issue. I've tested with multiple accounts, but never with multiple vehicles in a single account, so that didn't even cross my mind.

The HA integration fetches the current vehicle status every 3 minutes, but that requires a separate API call per vehicle. What I suspect is happening is that the two API calls for the two vehicles are being made at the exact same time. When the access token has expired and both calls need to re-authenticate, the server sees two simultaneous login attempts which would probably explain the various strange errors we are seeing.

I should be able to change the integration to only make one API call at at time. Let me see if I can make that change and get a test version of the integration for you to try and see if it resolves the problem.

**Comment by bdr99** [comment]
@AllPurposeBen Here's a version that only makes one call to the Mazda API at a time. Can you extract this into your `custom_components` folder, restart HA, and let me know if it resolves the issue for you?

[mazda_concurrent_request_fix.zip](https://github.com/bdr99/pymazda/files/6859564/mazda_concurrent_request_fix.zip)

**Comment by AllPurposeBen** [comment]
So far so good, but would like to give it a day or two to be sure.

This may be an unrelated issue but whilst I was looking at today's logs, [I saw a number of tracebacks in a row](https://gist.github.com/AllPurposeBen/7ddc5ecab75f2b52dec78a255faf71d1) that seem to indicate a connection issue with Mazda's servers. Happened for about an hour and then cleared on its own so maybe it was a transient issue on Mazda's end. FYI in case it's useful.

**Comment by bdr99** [comment]
I actually have the same entries in my HA logs at the same exact time of day as you. So it must have been an issue with the Mazda servers.

**Comment by AllPurposeBen** [comment]
Ran solid all weekend, no issues in the logs. I think we're good 😄

**Comment by bdr99** [comment]
@AllPurposeBen That's great! I'll open a PR to get these changes merged into HA. Thank again for the testing.

**Comment by AllPurposeBen** [comment]
Thank you for all your hard work on this!!

**Comment by bdr99** [comment]
The fix is now released in HA 2021.8, so I'm closing this issue.
